pkg/server/ring.go

```diff
@@ -0,0 +1,275 @@
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/storage/unified/resource"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health/grpc_health_v1"
+)
+
+const ringKey = "ring"
+const ringName = "unified_storage"
+const numTokens = 128
+const heartbeatTimeout = time.Minute
+
+var metricsPrefix = ringName + "_"
+
+func (ms *ModuleServer) initRing() (services.Service, error) {
+	if !ms.cfg.EnableSharding {
+		return nil, nil
+	}
+
+	if ms.cfg.MemberlistJoinMember == "" {
+		return nil, fmt.Errorf("bad sharding configuration. Missing MemberlistJoinMember")
+	}
+
+	if ms.cfg.RingListenPort == 0 {
+		return nil, fmt.Errorf("bad sharding configuration. Missing RingListenPort")
+	}
+
+	logger := log.New("resource-server-ring")
+	reg := prometheus.WrapRegistererWithPrefix(metricsPrefix, ms.registerer)
+
+	grpcclientcfg := &grpcclient.Config{}
+	flagext.DefaultValues(grpcclientcfg)
+	pool := newClientPool(*grpcclientcfg, logger, reg)
+
+	dnsProviderReg := prometheus.WrapRegistererWithPrefix(
+		metricsPrefix,
+		prometheus.WrapRegistererWith(
+			prometheus.Labels{"component": "memberlist"},
+			ms.registerer,
+		),
+	)
+	dnsProvider := dns.NewProvider(logger, dnsProviderReg, dns.GolangResolverType)
+
+	KVStore := kv.Config{Store: "memberlist"}
+
+	memberlistKVsvc := memberlist.NewKVInitService(toMemberlistConfig(ms.cfg), logger, dnsProvider, ms.registerer)
+	KVStore.MemberlistKV = memberlistKVsvc.GetMemberlistKV
+
+	ringStore, err := kv.NewClient(
+		KVStore,
+		ring.GetCodec(),
+		kv.RegistererWithKVName(reg, metricsPrefix),
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create KV store client: %s", err)
+	}
+
+	lifecyclerCfg, err := toLifecyclerConfig(ms.cfg, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize ruler's lifecycler config: %s", err)
+	}
+
+	// Define lifecycler delegates in reverse order (last to be called defined first because they're
+	// chained via "next delegate").
+	delegate := ring.BasicLifecyclerDelegate(ring.NewInstanceRegisterDelegate(ring.JOINING, numTokens))
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, logger)
+	delegate = ring.NewAutoForgetDelegate(heartbeatTimeout*2, delegate, logger)
+
+	lifecycler, err := ring.NewBasicLifecycler(
+		lifecyclerCfg,
+		ringName,
+		ringKey,
+		ringStore,
+		delegate,
+		logger,
+		reg,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize ruler's lifecycler: %s", err)
+	}
+
+	rulerRing, err := ring.NewWithStoreClientAndStrategy(
+		toRingConfig(ms.cfg, KVStore),
+		ringName,
+		ringKey,
+		ringStore,
+		ring.NewIgnoreUnhealthyInstancesReplicationStrategy(),
+		reg,
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize ruler's ring: %s", err)
+	}
+
+	startFn := func(ctx context.Context) error {
+		err := memberlistKVsvc.StartAsync(ctx)
```

Comment:
If we ever plan to add more rings, we should extract memberlistKV into its own service and init method, so that single memberlistKV can be shared between all the rings.